ntfsck: apply lost+found directory for orphaned inodes

jaysim opened the pull request:
If lost+found does not exist, ntfsck will create it.

And, if orphaned inode can't find proper parent or parent has been deleted, try to add to lost+found.

namjaejeon commented on src/ntfsck.c:
```diff
@@ -647,6 +665,119 @@ static int ntfsck_find_and_check_index(ntfs_inode *parent_ni, ntfs_inode *ni,
+
+	fn->parent_directory = MK_LE_MREF(lost_found->mft_no,
+			le16_to_cpu(lost_found->mrec->sequence_number));
+	ntfs_inode_mark_dirty(ni);
```
Can I know why you added inode dirty here ?

jaysim replied:
not ni, inode that has $FN, (ctx->ntfs_ino), I will remove it.

namjaejeon commented on src/ntfsck.c:
```diff
@@ -647,6 +665,119 @@ static int ntfsck_find_and_check_index(ntfs_inode *parent_ni, ntfs_inode *ni,
+	if (ret == STATUS_OK) { /* index already exist in parent inode */
+		printf("ntfsck_find_and_check_index %d\n", ret);
+		return STATUS_OK;
+	}
```
} else if {

}

namjaejeon commented on src/ntfsck.c:
```diff
@@ -1817,20 +1819,90 @@ static int ntfsck_truncate_attr(ntfs_attr *na, s64 new_size)
+		}
+	}
+
+#if 0
```
test code ? If this code is not needed, please remove it.

jaysim replied:
It removed in latest patchset

namjaejeon commented on src/ntfsck.c:
```diff
@@ -2613,6 +2567,54 @@ static int ntfsck_scan_index_entries_btree(ntfs_volume *vol)
+		return 1;
+	}
+
+	ni = ntfsck_check_root_inode(vol);
```
root inode should be checked in ntfsck_check_system_files()...

jaysim replied:
then it should check mft record bitmap for root inode in ntfsck_check_system_files().
I'll remove it. I think it's better to handle in new patchset

namjaejeon commented on src/ntfsck.c:
```diff
@@ -2002,118 +2002,83 @@ static int ntfsck_check_file(ntfs_inode *ni)
 
-
-		/* check size */
-		if ((ni->allocated_size != na->allocated_size) ||
```
you remove sizes checks, where are they checked ?

jaysim replied:
I thought it was not necessary because related size fields modified in ntfs_attr_update_mapping_pairs(), but that is only in case of corrupted attribute. 
so I'll restore it.

namjaejeon commented on src/ntfsck.c:
```diff
@@ -686,7 +699,17 @@ static int ntfsck_add_inode_to_parent(ntfs_volume *vol, ntfs_inode *parent_ni,
+	ret = ntfsck_find_and_check_index(parent_ni, ni, fn, FALSE);
+	if (ret == STATUS_OK) { /* index already exist in parent inode */
+		return STATUS_OK;
+	}
```
} else if () {

}

namjaejeon commented on src/ntfsck.c:
```diff
@@ -2526,6 +2590,29 @@ static int ntfsck_scan_index_entries_btree(ntfs_volume *vol)
+	if (lf_mftno == (u64)-1) {
+		/* create 'lost+found' directory */
+		ucs_namelen = ntfs_mbstoucs(FILENAME_LOST_FOUND, &ucs_name);
+		if (ucs_namelen != -1) {
```
```
WARNING: braces {} are not necessary for single statement blocks
#545: FILE: src/ntfsck.c:2598:
+		if (ucs_namelen != -1) {
+			lf_ni = ntfs_create(ni, 0, ucs_name, ucs_namelen, S_IFDIR);
+		}
```

namjaejeon commented on libntfs-3g/attrib.c:
```diff
@@ -131,10 +131,28 @@ s64 ntfs_get_attribute_value_length(const ATTR_RECORD *a)
+	}
+
+	a->value_length = cpu_to_le32(length);
+	return;
```
```
WARNING: void function return statements are not generally useful
#74: FILE: libntfs-3g/attrib.c:154:
+	return;
+}
```

namjaejeon commented on src/ntfsck.c:
```diff
@@ -1178,28 +1178,18 @@ static int ntfsck_check_file_name_attr(ntfs_inode *ni, FILE_NAME_ATTR *ie_fn,
-		}
-
-		if (idx_pdir != mft_pdir || idx_pdir_seq != mft_pdir_seq) {
+	if ((idx_pdir != mft_pdir) ||
```
```
CHECK: Unnecessary parentheses around 'idx_pdir != mft_pdir'
#98: FILE: src/ntfsck.c:1181:
+	if ((idx_pdir != mft_pdir) ||
+		(idx_pdir_seq != mft_pdir_seq) ||
+		(mft_pdir != ictx->ni->mft_no)) {

CHECK: Unnecessary parentheses around 'idx_pdir_seq != mft_pdir_seq'
#98: FILE: src/ntfsck.c:1181:
+	if ((idx_pdir != mft_pdir) ||
+		(idx_pdir_seq != mft_pdir_seq) ||
+		(mft_pdir != ictx->ni->mft_no)) {

CHECK: Unnecessary parentheses around 'mft_pdir != ictx->ni->mft_no'
#98: FILE: src/ntfsck.c:1181:
+	if ((idx_pdir != mft_pdir) ||
+		(idx_pdir_seq != mft_pdir_seq) ||
+		(mft_pdir != ictx->ni->mft_no)) {

CHECK: Alignment should match open parenthesis
#99: FILE: src/ntfsck.c:1182:
+	if ((idx_pdir != mft_pdir) ||
+		(idx_pdir_seq != mft_pdir_seq) ||
```

namjaejeon commented on src/ntfsck.c:
```diff
@@ -2129,41 +2119,196 @@ static int ntfsck_set_mft_record_bitmap(ntfs_inode *ni)
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {
+
```
```
CHECK: Blank lines aren't necessary after an open brace '{'
#477: FILE: src/ntfsck.c:2199:
+			(type != AT_LOGGED_UTILITY_STREAM)) {
+
```

namjaejeon commented on src/ntfsck.c:
```diff
@@ -2129,41 +2119,196 @@ static int ntfsck_set_mft_record_bitmap(ntfs_inode *ni)
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {
```
```
CHECK: Unnecessary parentheses around 'type != AT_STANDARD_INFORMATION'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_FILE_NAME'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_OBJECT_ID'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_SECURITY_DESCRIPTOR'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_VOLUME_NAME'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_VOLUME_INFORMATION'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_DATA'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_INDEX_ROOT'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_INDEX_ALLOCATION'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_BITMAP'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_REPARSE_POINT'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_EA_INFORMATION'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_EA'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_PROPERTY_SET'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {

CHECK: Unnecessary parentheses around 'type != AT_LOGGED_UTILITY_STREAM'
#462: FILE: src/ntfsck.c:2184:
+		if ((type != AT_STANDARD_INFORMATION) &&
+			(type != AT_FILE_NAME) &&
+			(type != AT_OBJECT_ID) &&
+			(type != AT_SECURITY_DESCRIPTOR) &&
+			(type != AT_VOLUME_NAME) &&
+			(type != AT_VOLUME_INFORMATION) &&
+			(type != AT_DATA) &&
+			(type != AT_INDEX_ROOT) &&
+			(type != AT_INDEX_ALLOCATION) &&
+			(type != AT_BITMAP) &&
+			(type != AT_REPARSE_POINT) &&
+			(type != AT_EA_INFORMATION) &&
+			(type != AT_EA) &&
+			(type != AT_PROPERTY_SET) &&
+			(type != AT_LOGGED_UTILITY_STREAM)) {
```

namjaejeon commented on libntfs-3g/inode.c:
```diff
@@ -717,6 +717,7 @@ int ntfs_inode_attach_all_extents(ntfs_inode *ni)
 	errno = 0;
 	ale = (ATTR_LIST_ENTRY *)ni->attr_list;
 	while ((u8*)ale < ni->attr_list + ni->attr_list_size) {
+
```
```
CHECK: Blank lines aren't necessary after an open brace '{'
#87: FILE: libntfs-3g/inode.c:720:
 	while ((u8*)ale < ni->attr_list + ni->attr_list_size) {
+
```

namjaejeon commented on libntfs-3g/runlist.c:
```diff
@@ -1019,12 +1020,27 @@ runlist_element *ntfs_mapping_pairs_decompress_i(const ntfs_volume *vol,
+			else
+				free(temp_rl);
+		}
+	}
```
} else {

}

namjaejeon commented on src/ntfsck.c:
```diff
@@ -1817,20 +1819,90 @@ static int ntfsck_truncate_attr(ntfs_attr *na, s64 new_size)
+				na->initialized_size);
+		ntfs_log_info(" inode: allocated_size(%ld), data_size(%ld) flags(%08x)\n",
+				ni->allocated_size, ni->data_size, ni->flags & (FILE_ATTR_SPARSE_FILE|FILE_ATTR_COMPRESSED));
+
```
```
CHECK: Blank lines aren't necessary before a close brace '}'
#336: FILE: src/ntfsck.c:1876:
+
+	}

```

namjaejeon commented on src/ntfsck.c:
```diff
@@ -1817,20 +1819,90 @@ static int ntfsck_truncate_attr(ntfs_attr *na, s64 new_size)
+#if 0
+	if (na->type != AT_DATA &&
+			((na->allocated_size != rls.data_size) ||
+			 (na->data_size != na->initialized_size))) {
```
```
CHECK: Unnecessary parentheses around 'na->allocated_size != rls.data_size'
#325: FILE: src/ntfsck.c:1865:
+	if (na->type != AT_DATA &&
+			((na->allocated_size != rls.data_size) ||
+			 (na->data_size != na->initialized_size))) {

CHECK: Unnecessary parentheses around 'na->data_size != na->initialized_size'
#325: FILE: src/ntfsck.c:1865:
+	if (na->type != AT_DATA &&
+			((na->allocated_size != rls.data_size) ||
+			 (na->data_size != na->initialized_size))) {
```

jaysim replied:
this #if 0 ~ #endif block was removed.

jaysim committed:
ntfsck: fix checking if system file is or not

Using utils_is_metadata() to distinguish inode is system file or not.
Previous that function do not check multiple depth system file like
$Extend/$RmMetadata/$TxfLog.
utils_is_metadata() is modified to distinguish above system files adding
$STD_INFO's flags checking.
(all system files have attribute flag for hidden and system)
And modifiy all place where check if file is metadata.

Add ntfsck_check_system_inode() function to check system files.

TODO: check system file more detail respecitively.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>
